AWS/aws/event_aggregator.py

```diff
@@ -0,0 +1,169 @@
+    ):
+        super().__init__()
+        self.event_aggregator = event_aggregator
+        self.f_must_stop = False
+        self.ttl = ttl
+        self.on_flush_func = on_flush_func
+        self.delay = delay
+
+    def stop(self) -> None:
+        self.f_must_stop = True
+
+    def run(self) -> None:
+        try:
+            while not self.f_must_stop:
+                time.sleep(self.delay)
+                for aggregated_event in self.event_aggregator.flush_all_ttl(ttl=self.ttl):
+                    self.on_flush_func(aggregated_event)
+        finally:
+            for aggregated_event in self.event_aggregator.flush_all():
+                self.on_flush_func(aggregated_event)
+
+
+class EventAggregator:
+    ttl_thread: EventAggregatorTTLThread | None
+    lock: threading.Lock
+
+    def __init__(self, aggregation_definitions: dict[str, list[Fingerprint]]):
+        self.aggregation_definitions = aggregation_definitions
+        self.aggregations: dict[str, Aggregation] = dict()
+        self.lock = threading.Lock()
+
+    def start_flush_on_ttl(self, ttl: int, on_flush_func: Callable[[dict], None], delay: float = 10):
+        """
+        This method MUST be called to trigger the execution of the ttl thread
+        """
+        self.ttl_thread = EventAggregatorTTLThread(
+            event_aggregator=self, ttl=ttl, on_flush_func=on_flush_func, delay=delay
+        )
+        self.ttl_thread.start()
+
+    def stop(self):
+        """
+        This method is only meant to stop the ttl thread
+        """
+        if self.ttl_thread:
+            self.ttl_thread.stop()
+            self.ttl_thread.join()
+
+    def flush_all(self) -> list[dict]:
+        """
+        returns all the aggregated events and remove them from the ongoing aggregations
+        """
+        aggregated_events = []
+
+        # we prevent any actions on the aggregations while flushing
+        with self.lock:
+            for aggregation in self.aggregations.values():
+                if aggregation.count > 0:
+                    aggregated_events.append(aggregation.get_aggregated_event())
+
+            self.aggregations = dict()
+        return aggregated_events
+
+    def flush_all_ttl(self, ttl: int) -> list[dict]:
+        """
+        Returns (and delete) all the events we aggregate for at least the ttl time
+        """
+        event_hashes_to_flush = []
+        aggregated_events = []
+        # we prevent any actions on the aggregations while flushing
+        with self.lock:
+            for event_hash, aggregation in self.aggregations.items():
+                if aggregation.start + datetime.timedelta(seconds=ttl) < datetime.datetime.utcnow():
+                    event_hashes_to_flush.append(event_hash)
+
+                    if aggregation.count > 0:
+                        aggregated_events.append(aggregation.get_aggregated_event())
+
+            for event_hash in event_hashes_to_flush:
+                del self.aggregations[event_hash]
```

Comment:
Could be done in the first iteration since we iterate on `.items()`

Reply:
ok I didn't know this method was compatible with dict edition. I will simplify accordingly.

Reply:
.items() doesnt seems compatible with inplace edition of dict. I used list(.keys()) instead